File: backend/app/inspiration_generator.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
import hashlib
import re

from pydantic import BaseModel, Field


class InspirationKind(StrEnum):
    MOTTO = "motto"
    FAMOUS_PEOPLE_QUOTE = "famous_people_quote"
    MANTRA = "mantra"
    AFFIRMATION = "affirmation"
    VALUE = "value"
    BELIEF = "belief"
    QUALITY = "quality"


DEFAULT_KINDS = [
    InspirationKind.MOTTO,
    InspirationKind.FAMOUS_PEOPLE_QUOTE,
    InspirationKind.MANTRA,
    InspirationKind.AFFIRMATION,
    InspirationKind.VALUE,
    InspirationKind.BELIEF,
    InspirationKind.QUALITY,
]


class InspirationGenerateRequest(BaseModel):
    prompt: str = Field(..., min_length=1, max_length=600)
    kinds: list[InspirationKind] | None = None
    count_per_kind: int = Field(3, ge=1, le=8)


class InspirationItem(BaseModel):
    kind: InspirationKind
    text: str
    author_name: str | None = None
    source_note: str | None = None


class InspirationGenerateResponse(BaseModel):
    prompt: str
    normalized_prompt: str
    items: list[InspirationItem]
    generated_by: str = "python-template-generator-v1"


@dataclass(frozen=True)
class QuoteSeed:
    text: str
    author_name: str


QUOTE_SEEDS = [
    QuoteSeed("Well done is better than well said.", "Benjamin Franklin"),
    QuoteSeed("Energy and persistence conquer all things.", "Benjamin Franklin"),
    QuoteSeed("Genius is one percent inspiration and ninety-nine percent perspiration.", "Thomas Edison"),
    QuoteSeed("To thine own self be true.", "William Shakespeare"),
    QuoteSeed("No great thing is created suddenly.", "Epictetus"),
    QuoteSeed("He who is brave is free.", "Seneca"),
]


STOP_WORDS = {
    "a",
    "an",
    "and",
    "are",
    "for",
    "from",
    "give",
    "help",
    "into",
    "make",
    "me",
    "my",
    "of",
    "on",
    "our",
    "please",
    "the",
    "to",
    "with",
}


TEMPLATES: dict[InspirationKind, list[str]] = {
    InspirationKind.MOTTO: [
        "Build {theme} with steady hands.",
        "{Theme}: honest work, visible progress.",
        "Make {theme} useful, durable, and kind.",
        "Small wins, strong {focus}.",
    ],
    InspirationKind.MANTRA: [
        "One clear step for {theme}.",
        "Return to the work; protect the signal.",
        "Calm focus, then momentum.",
        "Less noise, more {focus}.",
    ],
    InspirationKind.AFFIRMATION: [
        "I can turn {theme} into a concrete next step.",
        "I keep promises to the work and to myself.",
        "I am allowed to build steadily before it looks impressive.",
        "I learn, adjust, and continue.",
    ],
    InspirationKind.VALUE: [
        "Clarity before scale.",
        "Progress that users can trust.",
        "Craft in service of usefulness.",
        "Truthful momentum over performance.",
    ],
    InspirationKind.BELIEF: [
        "{Theme} grows through small decisions repeated well.",
        "Reliable systems are built by people who respect uncertainty.",
        "The next honest experiment is enough to move forward.",
        "Better tools can make better collaboration ordinary.",
    ],
    InspirationKind.QUALITY: [
        "Patient ambition",
        "Practical courage",
        "Careful momentum",
        "Generous rigor",
    ],
}


def generate_inspiration(request: InspirationGenerateRequest) -> InspirationGenerateResponse:
    prompt = request.prompt.strip()
    normalized_prompt = _normalize_prompt(prompt)
    theme = _theme_from_prompt(prompt)
    focus = _focus_word(prompt)
    seed = _stable_seed(prompt)
    kinds = request.kinds or DEFAULT_KINDS
    items: list[InspirationItem] = []

    for kind in kinds:
        for index in range(request.count_per_kind):
            if kind == InspirationKind.FAMOUS_PEOPLE_QUOTE:
                items.append(_quote_item(seed, index, theme))
                continue
            template = _pick(TEMPLATES[kind], seed, index)
            items.append(
                InspirationItem(
                    kind=kind,
                    text=template.format(theme=theme, Theme=theme.capitalize(), focus=focus),
                )
            )

    return InspirationGenerateResponse(
        prompt=prompt,
        normalized_prompt=normalized_prompt,
        items=items,
    )


def _quote_item(seed: int, index: int, theme: str) -> InspirationItem:
    quote = _pick(QUOTE_SEEDS, seed, index)
    return InspirationItem(
        kind=InspirationKind.FAMOUS_PEOPLE_QUOTE,
        text=quote.text,
        author_name=quote.author_name,
        source_note=f"Selected for: {theme}",
    )


def _pick[T](items: list[T], seed: int, index: int) -> T:
    return items[(seed + index) % len(items)]


def _stable_seed(prompt: str) -> int:
    digest = hashlib.sha256(prompt.strip().lower().encode("utf-8")).hexdigest()
    return int(digest[:8], 16)


def _normalize_prompt(prompt: str) -> str:
    normalized = re.sub(r"\s+", " ", prompt.strip().lower())
    return normalized or "motivation"


def _theme_from_prompt(prompt: str) -> str:
    words = _keywords(prompt)
    if not words:
        return "the work"
    return " ".join(words[:4])


def _focus_word(prompt: str) -> str:
    words = _keywords(prompt)
    return words[0] if words else "progress"


def _keywords(prompt: str) -> list[str]:
    words = re.findall(r"[A-Za-z0-9][A-Za-z0-9'-]*", prompt.lower())
    filtered = [word for word in words if word not in STOP_WORDS]
    return filtered or [word for word in words if word]

File: fe/utils/inspiration-generator.ts
import { API_BASE } from './api-config';

export type InspirationKind =
  | 'motto'
  | 'famous_people_quote'
  | 'mantra'
  | 'affirmation'
  | 'value'
  | 'belief'
  | 'quality';

export interface InspirationItem {
  kind: InspirationKind;
  text: string;
  author_name?: string | null;
  source_note?: string | null;
}

export interface InspirationGenerateRequest {
  prompt: string;
  kinds?: InspirationKind[];
  count_per_kind?: number;
}

export interface InspirationGenerateResponse {
  prompt: string;
  normalized_prompt: string;
  items: InspirationItem[];
  generated_by: string;
}

export const DEFAULT_INSPIRATION_KINDS: InspirationKind[] = [
  'motto',
  'famous_people_quote',
  'mantra',
  'affirmation',
  'value',
  'belief',
  'quality',
];

export const INSPIRATION_KIND_LABELS: Record<InspirationKind, string> = {
  motto: 'Mottos',
  famous_people_quote: 'Famous people quotes',
  mantra: 'Mantras',
  affirmation: 'Affirmations',
  value: 'Values',
  belief: 'Beliefs',
  quality: 'Qualities',
};

const STOP_WORDS = new Set([
  'a',
  'an',
  'and',
  'are',
  'for',
  'from',
  'give',
  'help',
  'into',
  'make',
  'me',
  'my',
  'of',
  'on',
  'our',
  'please',
  'the',
  'to',
  'with',
]);

const LOCAL_TEMPLATES: Record<Exclude<InspirationKind, 'famous_people_quote'>, string[]> = {
  motto: [
    'Build {theme} with steady hands.',
    '{Theme}: honest work, visible progress.',
    'Make {theme} useful, durable, and kind.',
  ],
  mantra: [
    'One clear step for {theme}.',
    'Return to the work; protect the signal.',
    'Calm focus, then momentum.',
  ],
  affirmation: [
    'I can turn {theme} into a concrete next step.',
    'I keep promises to the work and to myself.',
    'I learn, adjust, and continue.',
  ],
  value: [
    'Clarity before scale.',
    'Progress that users can trust.',
    'Craft in service of usefulness.',
  ],
  belief: [
    '{Theme} grows through small decisions repeated well.',
    'The next honest experiment is enough to move forward.',
    'Better tools can make better collaboration ordinary.',
  ],
  quality: ['Patient ambition', 'Practical courage', 'Careful momentum'],
};

const LOCAL_QUOTES = [
  { text: 'Well done is better than well said.', author_name: 'Benjamin Franklin' },
  { text: 'Energy and persistence conquer all things.', author_name: 'Benjamin Franklin' },
  {
    text: 'Genius is one percent inspiration and ninety-nine percent perspiration.',
    author_name: 'Thomas Edison',
  },
  { text: 'To thine own self be true.', author_name: 'William Shakespeare' },
  { text: 'No great thing is created suddenly.', author_name: 'Epictetus' },
  { text: 'He who is brave is free.', author_name: 'Seneca' },
];

export async function generateInspiration(
  request: InspirationGenerateRequest,
  endpoint = API_BASE,
): Promise<{ response: InspirationGenerateResponse; usedFallback: boolean }> {
  try {
    const res = await fetch(`${endpoint.replace(/\/$/, '')}/inspiration/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        prompt: request.prompt,
        kinds: request.kinds ?? DEFAULT_INSPIRATION_KINDS,
        count_per_kind: request.count_per_kind ?? 3,
      }),
    });
    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`Generator API error ${res.status}: ${text.slice(0, 240)}`);
    }
    return { response: await res.json(), usedFallback: false };
  } catch (error) {
    console.warn('Falling back to local inspiration generator', error);
    return { response: generateInspirationLocally(request), usedFallback: true };
  }
}

export function groupInspirationItems(items: InspirationItem[]) {
  return DEFAULT_INSPIRATION_KINDS.map((kind) => ({
    kind,
    label: INSPIRATION_KIND_LABELS[kind],
    items: items.filter((item) => item.kind === kind),
  })).filter((group) => group.items.length > 0);
}

function generateInspirationLocally(request: InspirationGenerateRequest): InspirationGenerateResponse {
  const prompt = request.prompt.trim();
  const normalized = prompt.toLowerCase().replace(/\s+/g, ' ');
  const seed = stableSeed(prompt);
  const theme = keywords(prompt).slice(0, 4).join(' ') || 'the work';
  const focus = keywords(prompt)[0] ?? 'progress';
  const kinds = request.kinds ?? DEFAULT_INSPIRATION_KINDS;
  const count = Math.min(Math.max(request.count_per_kind ?? 3, 1), 8);
  const items: InspirationItem[] = [];

  for (const kind of kinds) {
    for (let i = 0; i < count; i += 1) {
      if (kind === 'famous_people_quote') {
        const quote = pick(LOCAL_QUOTES, seed, i);
        items.push({ kind, ...quote, source_note: `Selected for: ${theme}` });
        continue;
      }
      const template = pick(LOCAL_TEMPLATES[kind], seed, i);
      items.push({
        kind,
        text: template
          .replaceAll('{theme}', theme)
          .replaceAll('{Theme}', capitalize(theme))
          .replaceAll('{focus}', focus),
      });
    }
  }

  return {
    prompt,
    normalized_prompt: normalized,
    items,
    generated_by: 'typescript-local-template-generator-v1',
  };
}

function pick<T>(items: T[], seed: number, index: number): T {
  return items[(seed + index) % items.length];
}

function stableSeed(value: string): number {
  let hash = 2166136261;
  for (let i = 0; i < value.length; i += 1) {
    hash ^= value.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return Math.abs(hash);
}

function keywords(prompt: string): string[] {
  const words = prompt.toLowerCase().match(/[a-z0-9][a-z0-9'-]*/g) ?? [];
  const filtered = words.filter((word) => !STOP_WORDS.has(word));
  return filtered.length ? filtered : words;
}

function capitalize(value: string): string {
  return value.length ? value[0].toUpperCase() + value.slice(1) : value;
}
